text/3199-unchecked-lifetime.md

```diff
@@ -0,0 +1,171 @@
+- The check that a reference does not outlive borrowed content is skipped when the borrowed content has `'?` lifetime. (This is the check that goes *unchecked*).
+- It cannot be used where a normal reference is expected.
+- assigning to or reading from `&'? T` and `&'? mut T` is unsafe
+
+Let's examine some of the implications of these features.
+
+It is safe in general to store into a value expecting `'?`, just like with raw pointers
+
+```rust
+struct A {
+    val: &'? usize
+}
+
+impl A {
+    fn store(&self, some_ref: &usize) {
+        self.val = some_ref; // assigning to val, not through val.
+    }
+}
+```
+An important but subtle point here is that it is only legal to have `&self` because of rule 2. Without it all references to A would outlive their borrowed content.
+
+The only way to use a type that has been stored with `&'?` is to use unsafe, or transmute it to a normal lifetime, which of course requires unsafe.
+```rust
+impl A {
+    // illegal
+    fn read_illegal(&self) -> usize {
+        *self.val // throws error due to rule 4
+    }
+
+    // legal, requiring unsafe.
+    fn read_deref(&self) -> usize {
+        unsafe { *self.val }
+    }
+
+    // legal, requiring unsafe.
+    fn read_transmute(&self) -> usize {
+        *(unsafe { core::mem::transmute::<&'? usize, &usize>(self.val) })
+    }
+}
+```
+note that read_transmute works for any type which is generic over a lifetime.
+
+Additionally, an important result of rule 3 is that methods defined on a reference to a type cannot be called on an unchecked reference unless they were specifically written to accept an unchecked reference, as unchecked references do not live long enough to be used where a normal reference would, like the method signature.
+```rust
+impl A {
+    // errors
+    fn get(&self) -> usize {
+        self.val.clone() // clone expects &self, so this errors as val does not live long enough.
+    }
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- One more pointer type is potentially confusing.
+- Potentially a trap for newer rust developers to just declare structs with unchecked lifetimes without realizing this is just as dangerous as throwing raw pointers around.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- There isn't any analogue in the higher type system to the lifetime erasure that occurs when coalescing from reference to pointer.
+- An alternative could be a macro-based library. This approach seems unlikely to check all the boxes though
+
+# Prior art
+[prior-art]: #prior-art
+
+- [Oroboros](https://docs.rs/ouroboros/0.13.0/ouroboros/attr.self_referencing.html) is a crate designed to facilitate self referential struct construction via macros. It is limited to references, and attempts to avoid unsafe.
+- [rental](https://docs.rs/rental/0.5.6/rental/), another macro based approach, is limited in a few ways and is somewhat clunky to use.
```

Comment:
Maybe add https://crates.io/crates/owning_ref here? It used be the "original" self-referential struct implementation as far as I know.